Don't ignore mypy errors by default

KapJI opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
### Motivation
Currently mypy checks only strictly typed parts of HA, everything else is just ignored.
This results that many components have type hints which are not checked and therefore completely useless.

### Solution
The idea is ignore errors in components which are currently broken and allow developers of these components to fix them and remove their component from ignored list. Even if they're not ready yet to enable strict checks.

This is while keeping components which are not broken in that state by checking their type hints.

If new component will have any type hints added they will be also checked. If new type hints are added, they should be checked, otherwise what's the point of having broken type hints?

This will allow to improve code by gradually fixing what is currently broken and doesn't satisfy its own type hints.

### Implementation
New `hassfest` plugin which generates and validates mypy config based on python list. Move mypy config from `setup.cfg` to `mypy.ini`.

Components which do not want to have strict checks enabled are now listed in `.no-strict-typing` file in the repo root which is easier to maintain. `hassfest` will use it for generating mypy config.

Strict checks are now enabled by default and this covers all core modules. New core modules will be automatically strictly checked.

### Current errors
Currently there are 3258 errors in 599 files in 238 components. Please note this is not with strict enabled, it's just checking existing type hints. With strict mode fully enabled there're 44k errors.

- [Full list of errors](https://gist.github.com/KapJI/f4eb10fb122834400e7120d94ddac9c7).
- [List of ignored components](https://gist.github.com/KapJI/75c3f59bdaa032c120d84590f49ed2d6). There are many generic components like `camera`, `config`, `climate`, `device_tracker`, `fan`, `recorder`, `script`, etc. They probably should be fixed first.

### Strict mode
I disabled `strict = true` because it's applied globally and it's an equivalent of `mypy --strict`. Instead I enabled the same checks manually. From `mypy --help`:
```
  --strict                  Strict mode; enables the following flags: --warn-unused-configs, --disallow-any-generics, --disallow-subclassing-any, --disallow-untyped-calls, --disallow-untyped-defs,
                            --disallow-incomplete-defs, --check-untyped-defs, --disallow-untyped-decorators, --no-implicit-optional, --warn-redundant-casts, --warn-unused-ignores, --warn-return-any, --no-
                            implicit-reexport, --strict-equality
```

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

KapJI committed:
Don't ignore mypy errors by default

bdraco commented:
@KapJI CI will fail due to https://github.com/home-assistant/core/pull/49271

frenck commented on setup.cfg:
```diff
@@ -35,18 +35,30 @@ noqa-require-code = True
 allow_any_generics = true
 implicit_reexport = true
+
+[mypy-homeassistant.components.adguard.*,homeassistant.components.aemet.*,homeassistant.components.airly.*,homeassistant.components.alarmdecoder.*,homeassistant.components.alexa.*,homeassistant.components.almond.*,homeassistant.components.amcrest.*,homeassistant.components.analytics.*,homeassistant.components.asuswrt.*,homeassistant.components.atag.*,homeassistant.components.aurora.*,homeassistan ...
```
This is not maintainable.

KapJI replied:
Move to `mypy.ini` or something else? What should be the input of hassfest?

marvin-w replied:
What's the point of this change if all components are still being ignored after creation?

If HA were to promote typing more other options should be considered (maybe adding it as part of the integration quality scale) - or are new integrations then forced to use typing?

KapJI replied:
Not all, only those which have type hints but they're broken. And they're broken because they were added but weren't actually checked. This will prevent this in the future because new type hints will be checked and allow gradually to fix broken ones. Also this enables checks for components which have type hints which are not broken (yet).

marvin-w replied:
Yep, should have read everything ;). Makes sense now.

frenck replied:
I think in this case generating a `mypy.ini` makes more sense, as generating the `setup.cfg` sounds something we shouldn't do either.

For the input, dunno, maybe a `.mypyignore` in the root?

KapJI committed:
Fix coronavirus component

KapJI committed:
Add hassfest plugin for generating mypy config

KapJI committed:
Intentionally trigger hassfest error

KapJI committed:
Revert intentional error

KapJI committed:
Add rituals_perfume_genie to ignored list

KapJI commented:
Ok, how about this one? 

Python lists should be much easier to maintain and it doesn't require putting extra file in the repo root.
Also this keeps all mypy configuration in the single place.

KapJI commented:
@frenck what do you think?

KapJI committed:
Ignore profiler component

KapJI commented:
@frenck friendly ping. New broken type annotation are added almost every day, let's secure the current state at least.

@cdce8p I think you also may be interested in this.

KapJI commented:
@MartinHjelmare maybe you have any feedback on this?

MartinHjelmare commented on homeassistant/components/coronavirus/config_flow.py:
```diff
@@ -21,20 +22,21 @@ class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
-                coordinator.data.values(), key=lambda case: case.country
-            ):
-                self._options[case.country] = case.country
+            if coordinator.data is not None:
```
Should we group this condition with the one on line 31 so that we abort if we don't have data?

MartinHjelmare commented on script/hassfest/mypy_config.py:
```diff
@@ -0,0 +1,408 @@
+from .model import Config, Integration
+
+# Strictly typed parts of HA.
+# If your component is fully typed, please add it here to enable strict checks.
```
Another approach would be to add an item in the manifest.json that would populate the corresponding list. Similar to how codeowners is generated.

Then we'd just have the ignore list here. New integrations should never end up on the ignore list.

KapJI replied:
We can do this for components and add new field like `strictly_typed: true` in `manifest.json` but core modules don't have `manifest.json` so part of this list will still be needed.

But I think we can enable strict checks by default and then disable them for `components.*` and then enable them again per component basis. This way new modules in the core won't need to be added here and large part of this list can be removed.

MartinHjelmare replied:
I think we should not add the typing_ignored item to the manifest since we should be working to remove that case completely.

> But I think we can enable strict checks by default and then disable them for components.* and then enable them again per component basis where needed.

Do you mean that we should do this and add the `strictly_typed` item to the manifest?

KapJI replied:
Yes:
1. Enable strict check by default, this will be applied to all core modules.
2. Disable strict checks for `components.*`
3. Generate list of strictly typed components based on `strictly_typed` in their `manifest.json`.
4. Generate list of ingored components based on list in this hassfest plugin.

MartinHjelmare replied:
Sounds good! 👍

KapJI replied:
> I think we should not add the typing_ignored item to the manifest since we should be working to remove that case completely.

My idea is that can create more incentive for component developers to fix these errors quicker if they see it in their `manifest.json`.

KapJI committed:
Add strictly_typed to manifest.json

KapJI committed:
Update mypy_config plugin

KapJI commented:
Now it generates `mypy.ini` based on new field in `manifest.json`.

Because of that I removed these modules from strictly typed list, the whole component should be strictly typed to enable strict checks:
```
homeassistant.components.recorder.purge
homeassistant.components.recorder.repack
homeassistant.components.sonos.media_player
```

Also I added check that component which is marked with `strictly_typed` is not in ignored list, otherwise ignored list will override strict checks.

KapJI committed:
Make codecov happy

frenck commented:
This should not be part of the manifest. There is nothing at runtime that needs this.

KapJI commented:
@frenck what do you think about previous version?

@MartinHjelmare your opinion?

MartinHjelmare commented:
I think it can be part of the manifest. We have codeowners already there which is a similar problem.

KapJI commented:
I think having this is the manifest and in documentation will motivate developers to add type hints to new integrations since the beginning. And existing integrations are more likely to be updated.

Otherwise they need to figure out how this config is generated and update hassfest plugin which is in general more difficult and less clear step.

KapJI committed:
Add motioneye to ignored list

frenck commented:
There is really not need to have this in the manifest. The manifest is loaded at runtime as well, and typing has 0 addition in this. Neither do custom components have anything to do with this either (to which manifests applies as well).

I rather see a `.mypyignore` or `.typingignore` file (similar to `.coveragerc`).

I really do not approve/agree with the manifest being used for this.

KapJI committed:
Revert "Add strictly_typed to manifest.json"

This reverts commit 1e23129dee98332bd16f1aac0b0425507d5a6fe4.

KapJI committed:
List sctricly typed components in .typingstrict

KapJI commented:
Ok, strictly typed modules (only from components, strict checks for core modules are now enabled by default) are now listed in `.typingstrict` in the repo root which is more easily maintaintable than old super long line in `setup.cfg`.

Ignored modules are listed directly in hassfest plugin as before. Nothing new should be added there, only removed.

MartinHjelmare reviewed (approved):
Looks good!

KapJI committed:
Add fritzbox to ignored list

KapJI commented:
@frenck is this something you can approve?

frenck reviewed (changes requested):
This logic should be reverted.

Strict typing should be the default, if one doesn't want to do that, it should be listed. This to encourage the use of typing.

Good behavior / best practice should be default, exceptions for that need configuration and not the other way around.

KapJI committed:
Inverse logic

KapJI commented:
Ok, now there is `.no-strict-typing` file which have 950 components (out of 990) which have strict checks disabled.

KapJI commented:
@MartinHjelmare are you still happy with this?

KapJI committed:
Add fritz to ignored list

KapJI commented:
@cdce8p in `homeassistant/components/fritz/device_tracker.py` annotation was `Dict[str, any]` and pylint said nothing about `any`. I think this should be warning.

cdce8p commented:
> @cdce8p In `homeassistant/components/fritz/device_tracker.py` annotation was `Dict[str, any]` and pylint said nothing about `any`. I think this should be a warning.
> 
> Mypy could probably catch it but it's disabled for this component. Not sure if it worths adding in pylint.

IMO this is not something pylint should handle. I have no intention of turning it into a type checker 😅
You're right btw, mypy would report an error here: `Function "builtins.any" is not a valid type`

cdce8p commented:
@KapJI Just FYI, it would be better to open a new PR to fix pylint. That's something which can / should be merged fairly quickly and isn't really related to the original topic. For now there is already one: #49655

KapJI commented:
@cdce8p Yes, I have no intention to fix it here, it was just to demonstrate that CI passes. I'll rebase and that change will go away.

KapJI commented:
@frenck is this what you meant?

frenck reviewed (approved):
Yeah, awesome!

Now, we can start reducing this list and prevent (well, encourage) new entries from being added 👍

KapJI commented:
I rebased to current `dev` and mypy passes, should be safe to merge it now.

I'd suggest to be careful with merging PRs which are not rebased on top of this as they can easily add broken type hints.

frenck commented:
> I'd suggest to be careful with merging PRs which are not rebased on top of this as they can easily add broken type hints.

Those will cause build failures, which can be annoying, but not problematic :)

frenck commented:
Awesome work @KapJI! 🎉

I've adjusted our developer documentation to provide information around this as well.

Let's hope we'll see more nicely typed things from this! (Oh... and as it seems my integrations are in the naughty list 😢)

balloob commented:
We should not make new integrations require opt-out of type checking. This severely increases the barrier of entry to submit integrations to Home Assistant. 

I was helping someone today. They had made a PyPI package and scaffolded an integration. Pre-commit failed because a) new functions he added didn't have the typing and b) their PyPI package didn't have types and got called from a typed context. The mypy output is not straight forward to fix, and without the ability to commit, they can't even push up their work to ask for help (they didn't know about `--no-verify`).

This move will cause more integrations to go towards HACS, bypassing our review completely. If setting up people for a technically perfect contribution means we don't get contributions… we miss our goal. Having people go to HACS is worse as our reviewers catch a lot + we add more checks all the time that improve code quality of existing integrations.

We're open source and we need to keep the environment welcoming to attract new developers. Allowing for opt-in type checking allows new developers to grow as they learn more. We should not expect new developers to play the game at expert level out of the gate.

frenck commented:
I am a believer of default best practices, which including typing.

The only thing I cannot understand is how this is any different compare to testing? We require those by default, or else coverage will fail. Is that wrong too?

bdraco commented:
How about changing the scaffolding to ask them if they want to opt-in?

balloob commented:
We don't require tests for most part of the integration. We only require it for config flow because we have deemed it so important that adding that burden was worth it. However, we did make it as easy as possible to have all the required tests be generated based on the integration that is scaffolded. 

If you have incorrect types you can't commit your code. That is very frustrating. And the problem is that these are contributors that we won't hear about, because they quit.

A better approach is to make it part of the [integration quality scale](https://developers.home-assistant.io/docs/integration_quality_scale_index). I suggest we make it part of the gold level, which also introduces requirements for testing.

Danielhiversen commented:
Should we (always) require that the underlying library supports typing?
https://github.com/home-assistant/core/pull/49712
(I hope not, or maybe only for gold level)

balloob commented:
I don't think that Home Assistant should reject dependencies for lack of typing.

We should all not forget where we came from. Being a new programmer, playing with Python, trying to control your home with a couple of key strokes. We shouldn't try to just attract most senior Python programmers that have been exposed to typing at their work. We're open source and we should focus on being accessible to even beginning Python programmers. 

When we help a person learn Python, the odds that they stay around are a lot higher than some senior dev passing by to do a contribution. We have and had many core contributors that learned Python and/or working on big projects because of Home Assistant. Those people are amazing.

(Note, discussion is about integrations, not the core) 

Let's make type checking opt-in for new integrations and let's add it as a requirement of the gold level of the integration quality scale.

frenck commented:
> We don't require tests for most part of the integration

That is no different in this case. We indeed don't require it, but the CI will fail, unless you explicitly add stuff to coveragerc.

This method is no different in that regard.

> If you have incorrect types you can't commit your code. That is very frustrating.

That I agree on! But, isn't that solved by giving a proper error message? with e.g. a friendly instruction that says, either add typing or add it to the list?

The list has a additional big a advantage; it provides a to-do list (instead of a done list). Ok can be slimmed down.

frenck commented:
> Should we (always) require that the underlying library supports typing?
#49712
(I hope not, or maybe only for gold level)

We cannot do that, but we could require it maybe for certain QA scales. Same as with async or test requirements. 

However, adding typing to an integration when upstream package doesn't provide it, is hard in general an not and limition of anything Home Assistant. You'll probably need to cast and ignore a lot to get something like that even remotely correct or helpful.

balloob commented:
The problem with the error message is that it's raised by `pre-commit` and so disconnected from scaffolding/hassfest.

KapJI commented:
There are several difficulties with linking typing and gold level:

1. Some existing integrations with gold level may not pass it, we will needed to downgrade them unless they can be quickly fixed.
2. It will be harder or not possible to enable strict checks if integration doesn't satisfy other requirements for gold level.

Regarding pre-commit: it's possible to run mypy from some wrapper which will check if integration is added in config and will provide clear instructions how to add it there if person is not ready yet to fix these errors. I've seen something similar in urllib3.

Also it can show warning if integration is excluded from type checking at all (added to IGNORED_LIST). This should help to get these integrations with broken type annotations to be fixed sooner.

And about friendliness to newcomers. From my experience you can attract more people to work on the project if you just try to be more nice and polite when people are sending their contributions. For me friendly environment where my contributions are welcomed is much bigger factor for my desire to contribute more than solving technical roadblocks like adding an extra line to some config. I really believe you can do this.

frenck commented:
> It will be harder or not possible to enable strict checks if integration doesn't satisfy other requirements for gold level.

This should not be coupled with the "qa_scale" manifest status for deciding to check yes/no. Even an integration without any rating (for whatever reason), should just be able to opt-in/out regardless of the scale.

QA scale is a manual process at this point.
